src/js/mqtt.js

```diff
@@ -0,0 +1,308 @@
+/* Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var net = require('net');
+var util = require('util');
+var EventEmitter = require('events').EventEmitter;
+
+util.inherits(MQTTClient, EventEmitter);
+
+PacketTypeEnum = {
+  PUBACK : 4,
+  PUBREC : 5,
+  PUBREL : 6,
+  PUBCOMP : 7
+};
+
+function MQTTClient(options) {
+  if (!(this instanceof MQTTClient)) {
+    return new MQTTClient(options);
+  }
+
+  EventEmitter.call(this);
+
+  this._protocol = options.protocol || 'net';
+
+  this._clientOptions = Object.create(options, {
+    host: { value: options.host || '127.0.0.1'},
+    port: { value: options.port || 8883 },
+    qos: { value: options.qos || 0 },
+    keepalive: { value: options.keepalive || 60 },
+  });
+
+  this._socket = new net.Socket();
+  this._socket.on('error', this.onerror);
+
+  this._isConnected = false;
+  this._reconnecting = false;
+  this._package_id = 0;
+}
+
+/*
+ * Connect to an MQTT broker.
+ */
+function MqttConnect(socket, options) {
+  var buff = native.connect(options);
+  socket.write(buff);
+}
+
+MQTTClient.prototype.connect = function(callback) {
+  this._clientOptions.cb = callback;
+  var jsref = this;
+
+  if (this._protocol === 'tls') {
+    var tls = require('tls');
+    this._socket = tls.connect(this._clientOptions);
+    this._socket.on('secureConnect', function() {
+      MqttConnect(this, jsref._clientOptions);
+    });
+  } else {
+    this._socket = net.connect(this._clientOptions);
+    this._socket.on('connect', function() {
+      MqttConnect(this, jsref._clientOptions);
+    });
+  }
+
+  if (util.isFunction(callback)) {
+    this.on('connect', callback);
+  }
+
+  this._socket.on('data', function(data) {
+    jsref.ondata(jsref, data);
+  });
+  this._socket.on('error', function(e) {
+    jsref.emit('error', e);
+  });
+  this._socket.on('end', this.ondisconnect);
+};
+
+MQTTClient.prototype.disconnect = function(error) {
```

Comment:
I think `end` is more common than `disconnect`.

Reply:
There actually is a packet type which has the name `DISCONNECT`, causing the server to properly disconnect the client.